Collection of GUC parameters after benchmarking

kuzeyardic opened the pull request.

Commit:
feat: Add Citus distributed table support to PgDiskANN client

This commit integrates Citus distributed table functionality with the PgDiskANN
vector database client to enable horizontal scaling for large vector datasets.

Key Changes:
- Enhanced table creation to support Citus distributed tables
- Added automatic shard distribution optimized for 500K rows with 1536 dimensions
- Implemented intelligent fallback to regular tables when Citus is unavailable
- Added configuration options for Citus distribution control

Technical Details:
1. Modified _create_table() to call _create_distributed_table() after base table creation
2. Added _create_distributed_table() method with:
   - Citus extension detection
   - Hash distribution on 'id' column for even data distribution
   - Configurable shard count (default: 8 shards for optimal 500K row performance)
   - Comprehensive logging for distribution status

3. Enhanced connection method to enable required extensions:
   - pg_diskann: DiskANN vector indexing
   - citus: Distributed table functionality
   - vector: Vector data type support

4. Added configuration parameters:
   - enable_citus_distribution: Toggle distributed table creation
   - shard_count: Configurable shard count for performance tuning

5. Extended CLI interface with new options:
   - --enable-citus-distribution/--disable-citus-distribution
   - --shard-count (default: 8)

Performance Benefits:
- Distributes 500K vectors across 8 shards (~62.5K rows per shard)
- Enables parallel DiskANN index creation across multiple nodes
- Maintains DiskANN's approximate nearest neighbor search performance
- Provides horizontal scalability for large vector datasets

The implementation ensures backward compatibility with existing setups while
enabling advanced distributed functionality for production workloads.

Commit:
feat: Add Citus distributed table support to PgDiskANN

- Enhanced PgDiskANN with Citus integration for horizontal scaling
- Added smart extension management for Azure PostgreSQL compatibility
- Implemented distributed table creation with configurable shard count
- Added comprehensive logging and error handling
- Updated .gitignore for documentation and log files
- Achieved 10.7% faster data loading with distributed tables
- Maintained perfect accuracy (1.0 recall) across configurations

Commit:
chore: Remove pgdiskann_results.csv file

- Removed CSV results file to clean up repository
- All benchmark results are now properly excluded via .gitignore
- JSON results in results/ directory are automatically ignored
- Ready for new experiments and benchmarks

Commit:
refactor: Simplify .gitignore file

- Removed specific documentation file exclusions
- Kept essential exclusions: results/, logs/, config files
- Simplified structure for cleaner repository management
- Additional log files still excluded via pattern matching

Commit:
Add analysis scripts and results to .gitignore

- Added specialized_analysis.py to .gitignore for local analysis tool
- Added analyses/ directory to .gitignore for generated analysis results
- Keeps analysis tools local while maintaining clean repository

Commit:
feat: Enhanced privacy protection for benchmark experiments

- Updated db_label to 'Private-Test-Setup-A' for anonymous testing
- Improved .gitignore to protect sensitive configuration and result files
- Added comprehensive privacy protection for PgDiskANN experiments
- Documented db_label functionality and privacy implications
- Ensured local-only result storage without automatic public sharing

Commit:
guc collection complete

onurctirtir commented on vectordb_bench/backend/clients/pgdiskann/pgdiskann.py:
```diff
@@ -263,16 +293,97 @@ def _create_table(self, dim: int):
+                
+                # Create distributed table with hash distribution on id column
+                self.cursor.execute(
+                    sql.SQL("SELECT create_distributed_table({table_name}, 'id');").format(
```
if / when we really want to set shard_count, we should provide it as a parameter to create_distributed_table (create_distributed_table supports "shard_count" parameter)

onurctirtir commented on vectordb_bench/backend/clients/pgdiskann/pgdiskann.py:
```diff
@@ -263,16 +293,97 @@ def _create_table(self, dim: int):
+                    # Get worker node information
+                    self.cursor.execute("SELECT nodename, nodeport FROM pg_dist_node WHERE isactive = true;")
+                    worker_info = self.cursor.fetchall()
+                    log.info(f"{self.name} Active worker nodes: {worker_info}")
```
should remove these for now

onurctirtir commented on vectordb_bench/backend/clients/pgdiskann/pgdiskann.py:
```diff
@@ -327,3 +438,371 @@ def search_embedding(
+                    # 1. Current GUC parameters (live state) - ACTIVE
+                    config_metrics['current_guc_parameters'] = self._collect_live_guc_parameters(cursor)
+                    
+                    # 2. Current Citus state (if enabled) - ACTIVE
```
can remove commented out code

onurctirtir commented on vectordb_bench/backend/clients/pgdiskann/pgdiskann.py:
```diff
@@ -327,3 +438,371 @@ def search_embedding(
+            guc_params = {}
+            
+            # Define all the Citus parameters to collect
+            citus_parameters = [
```
we only need these GUCs

diskann.l_value_is
diskann.iterative_search

citus.max_adaptive_executor_pool_size
citus.max_cached_connection_lifetime
citus.max_cached_conns_per_worker
citus.max_client_connections
citus.max_shared_pool_size
citus.local_shared_pool_size
citus.executor_slow_start_interval
citus.force_max_query_parallelization
citus.enable_binary_protocol
citus.enable_repartition_joins

onurctirtir commented on vectordb_bench/backend/clients/pgdiskann/pgdiskann.py:
can use sth like this to get diskann index opts like max_neighbors and l_value_ib

```sql
select reloptions
from pg_index
join pg_class on (oid = indexrelid)
where relname = 'pgdiskann_index'
```

onurctirtir commented on vectordb_bench/backend/clients/pgdiskann/pgdiskann.py:
```diff
@@ -327,3 +438,371 @@ def search_embedding(
+            log.error(f"Error collecting Citus GUC parameters: {e}")
+            return {'error': str(e)}
+    
+    def _collect_citus_runtime_state(self, cursor) -> dict:
```
can just use this

```sql
select nodename, nodeport, count(*) as shard_count_on_node
from pg_dist_shard
join pg_dist_placement using (shardid)
join pg_class on (logicalrelid = oid)
join pg_dist_node using (groupid)
where relname = 'pg_diskann_collection'
group by (nodename, nodeport)
order by nodeport;
```

Commit:
guc parameters reduced, reviews applied, shard count setting works, diskann parameter connection will be finished not ready.

Commit:
feat: Add DiskANN parameter collection with session setup

- Implement proper DiskANN GUC parameter collection in post-benchmark analysis
- Add session parameter setup before querying diskann.l_value_is and diskann.iterative_search
- Set diskann.l_value_is from config and diskann.iterative_search to default 'Relaxed_Order'
- Achieve 100% success rate for both DiskANN parameters in collection
- Enhance logging and metadata tracking for DiskANN parameter collection
- Fix session-scoped parameter issue by replicating session setup from init() method

Commit:
Fix Citus transaction conflicts in post-benchmark analysis

- Add isolated connection for EXPLAIN query execution to avoid transaction conflicts
- Implement _collect_query_plans() method with dedicated connection and autocommit
- Fix simple_json_reader.py to show all list items without truncation
- Resolve 'cannot execute command because a local execution has accessed a placement' errors
- Enable comprehensive post-benchmark query plan collection for performance analysis

Technical changes:
- Use separate psycopg connection for EXPLAIN queries to isolate from main transaction
- Set citus.enable_local_execution=OFF before running EXPLAIN commands
- Add proper session parameter setup for query plan collection
- Improve error handling and logging for debugging transaction issues
